qutip/qip/circuit.py

```diff
@@ -848,6 +1122,101 @@ def _basis_SQRTISWAP(self, qc_temp, temp_resolved):
+        after adding all the gates and measurements on the circuit and returns
+        the resultant state after evolution
+        Parameters
+        ----------
+        state : ket
+                state to be observed on specified by density matrix.
+        cbits : List of ints
+                initialization of the classical bits
+
+        Returns
+        -------
+        state : returns the ket of the output state after running the circuit.
+        '''
+
+        if len(cbits) == self.num_cbits and cbits:
+            self.cbits = cbits
+        else:
+            self.cbits = [0] * self.num_cbits
+
+        if state.shape[0] != 2 ** self.N:
+            return ValueError("dimension of state is incorrect")
+
+        self.U_list = self.propagators()
+
+        ulistindex = 0
+
+        for operation in self.gates_and_measurements:
+
+            if isinstance(operation, Measurement):
+
+                probabilities, states = operation.measurement_comp_basis(state)
+                i = np.random.choice([0, 1],
+                                p=[probabilities[0], 1 - probabilities[0]])
+                state = states[i]
+                if operation.classical_store is not None:
+                    self.cbits[operation.classical_store] = i
+
+            else:
+
+                if (operation.classical_controls):
+                    apply_gate = all([self.cbits[i]
+                                for i in operation.classical_controls])
+                    if apply_gate:
+                        state = self.U_list[ulistindex] * state
+                        ulistindex += 1
+                    else:
+                        ulistindex += 1
+                        continue
+                else:
+                    state = self.U_list[ulistindex] * state
+                    ulistindex += 1
+
+        return state
+
+    def run_statistics(self, state, cbits=[], num_runs=1024):
+        '''
+        This is the circuit run function for num_runs run, must be called after
+        adding all the gates and measurements on the circuit and returns the
+        frequency with which each output state is observed after num_runs runs.
+
+        Parameters
+        ----------
+        state : ket
+                state to be observed on specified by density matrix.
+        cbits : List of ints
+                initialization of the classical bits.
+        num_runs : int
+                number of times the circuit is run with the specified input.
+
+        Returns
+        -------
+        results : List of [ket, int]
+                returns the number of times each state is observed as
+                output state on running the circuit num_runs times.
+        '''
+
+        state_freq = defaultdict(int)
+        tuple_to_state = defaultdict()
+
+        for i in range(num_runs):
+            final_state = self.run(state, cbits=[])
+            state_freq[tuple([complex(a) for a in final_state.full()])] += 1
```

Comment:
This seems to be quite vulnerable to floating-point error and global phase. The ideal way would be to check if the fidelity of two states is smaller than some threshold.

Reply:
Good idea, will do ! I guess the state output can just be the first instance (I'm assuming it doesn't matter much what the small deviations are)